src/jit/lowerxarch.cpp

```diff
@@ -805,6 +805,77 @@ void Lowering::LowerHWIntrinsicCC(GenTreeHWIntrinsic* node, NamedIntrinsic newIn
     }
 }
 
+//----------------------------------------------------------------------------------------------
+// LowerFusedMultiplyAdd: Changes NI_FMA_MultiplyAddScalar produced by Math(F).FusedMultiplyAdd
+//     to a better FMA intrinsics if there are GT_NEG around in order to eliminate them.
+//
+//  Arguments:
+//     node - The hardware intrinsic node
+//
+//  Notes:
+//     Math(F).FusedMultiplyAdd is expanded into NI_FMA_MultiplyAddScalar and
+//     depending on additional GT_NEG nodes around it can be:
+//
+//      x *  y + z -> NI_FMA_MultiplyAddScalar
+//      x * -y + z -> NI_FMA_MultiplyAddNegatedScalar
+//     -x *  y + z -> NI_FMA_MultiplyAddNegatedScalar
+//     -x * -y + z -> NI_FMA_MultiplyAddScalar
+//      x *  y - z -> NI_FMA_MultiplySubtractScalar
+//      x * -y - z -> NI_FMA_MultiplySubtractNegatedScalar
+//     -x *  y - z -> NI_FMA_MultiplySubtractNegatedScalar
+//     -x * -y - z -> NI_FMA_MultiplySubtractScalar
+//
+void Lowering::LowerFusedMultiplyAdd(GenTreeHWIntrinsic* node)
+{
+    GenTreeArgList*     argList = node->gtGetOp1()->AsArgList();
+    GenTreeHWIntrinsic* createScalarOps[4]; // 3 arguments with 1-based index
+
+    for (int i = 1; i < 4; i++)
+    {
+        assert(argList->Current() != nullptr);
+        if (argList->Current()->OperIsHWIntrinsic())
+        {
+            GenTreeHWIntrinsic* hwArg = argList->Current()->AsHWIntrinsic();
+            if (hwArg->gtHWIntrinsicId == NI_Vector128_CreateScalarUnsafe)
+            {
+                createScalarOps[i] = hwArg;
+                argList            = argList->Rest();
+                continue;
+            }
+        }
+
+        return; // Math(F).FusedMultiplyAdd is expected to emit three NI_Vector128_CreateScalarUnsafe
+                // but it's also possible to use NI_FMA_MultiplyAddScalar directly
+    }
+    assert(argList == nullptr);
+
+    GenTree* argX = createScalarOps[1]->gtGetOp1();
+    GenTree* argY = createScalarOps[2]->gtGetOp1();
+    GenTree* argZ = createScalarOps[3]->gtGetOp1();
+
+    const bool negMul = argX->OperIs(GT_NEG) != argY->OperIs(GT_NEG);
```

Comment:
Will the JIT have optimized away double negations by this point?

e.g. The user has:
```csharp
MultiplyAddNegated(-5, y, z);

float MultiplyAddNegated(float x, float y, float z)
{
    return MathF.MultiplyAdd(-x, y, z);
}
```

I don't think this is something we should handle here if not; it's more just a sanity check on my end on if this will be handled somewhere 😄

Reply:
Heh, I actually wanted to file an issue for it yesterday (and even had a simple fix - to remove `GT_NEG->GT_NEG->x`, the fix had 0 jit-diff results but probably still worth implementing). Let me file an issue.